RecoDataProducts/src/IntensityInfoCalo.cc

```diff
@@ -0,0 +1,41 @@
+// Implementation of methods
+#include "Offline/RecoDataProducts/inc/IntensityInfoCalo.hh"
+#include <algorithm>
+
+namespace mu2e {
+
+  // Store a Caphri hit
+  bool IntensityInfoCalo::addCaphriHit(const double energy, const int ID) {
+
+    // Map the crystal ID to an index between 0-3
+    const auto itr = std::find(CaloConst::_caphriId.begin(), CaloConst::_caphriId.end(), ID);
+    if(itr == CaloConst::_caphriId.end()) return false; // not a CAPHRI crystal
+    const unsigned short index = std::distance(CaloConst::_caphriId.begin(), itr);
+    if(index > 3) return false; // must be between 0-3
+
+    const unsigned short e_short = energy / caphriEnergyUnits_; // store in different units from MeV input to adjust the precision
+
+    // encode the energy + index into a single short
+    const unsigned short hit_info = e_short | index << caphriIndexBits_;
+
+    // add it to the hit vector
+    caphriHits_.push_back(hit_info);
+
+    return true;
+  }
+
+  // Decode CAPHRI hit information
+  void IntensityInfoCalo::getCaphriHit(const size_t ihit, double& energy, int& ID) const {
+    ID = -1; energy = 0.;
+    if(ihit >= caphriHits_.size()) return;
```

Comment:
Is there a legitimate use case for requesting a non-existing hit? If not I suggest treating OOB as an error (exception). That also avoids having to return a negative ID, which otherwise is unsigned.

Reply:
I was unsure of where to put the error throwing initially. I've updated the product to throw errors now, and removed this error checking from the modules